src/core/dense_set.h

```diff
@@ -0,0 +1,471 @@
+
+ private:
+  struct DenseLinkKey;
+  // we can assume that high 12 bits of user address space
+  // can be used for tagging. At most 52 bits of address are reserved for
+  // some configurations, and usually it's 48 bits.
+  // https://www.kernel.org/doc/html/latest/arm64/memory.html
+  static constexpr size_t kLinkBit = 1ULL << 52;
+  static constexpr size_t kDisplaceBit = 1ULL << 53;
+  static constexpr size_t kDisplaceDirectionBit = 1ULL << 54;
+  static constexpr size_t kTagMask = 4095ULL << 51;  // we reserve 12 high bits.
+
+  struct DensePtr {
+    explicit DensePtr(void* p = nullptr) : ptr_(p) {
+    }
+
+    bool IsObject() const {
+      return (uintptr_t(ptr_) & kLinkBit) == 0;
+    }
+
+    bool IsLink() const {
+      return (uintptr_t(ptr_) & kLinkBit) == kLinkBit;
+    }
+
+    bool IsEmpty() const {
+      return ptr_ == nullptr;
+    }
+
+    void* Raw() const {
+      return (void*)(uintptr_t(ptr_) & ~kTagMask);
+    }
+
+    bool IsDisplaced() const {
+      return (uintptr_t(ptr_) & kDisplaceBit) == kDisplaceBit;
+    }
+
+    void SetLink(DenseLinkKey* lk) {
+      ptr_ = (void*)(uintptr_t(lk) | kLinkBit);
+    }
+
+    void SetDisplaced(int direction) {
+      ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceBit);
+      if (direction == 1) {
+        ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceDirectionBit);
+      }
+    }
+
+    void ClearDisplaced() {
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceBit);
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceDirectionBit);
+    }
+
+    // returns 1 if the displaced node is right of the correct bucket and -1 if it is left
+    int GetDisplacedDirection() const {
+      return (uintptr_t(ptr_) & kDisplaceDirectionBit) == kDisplaceDirectionBit ? 1 : -1;
+    }
+
+    void Reset() {
+      ptr_ = nullptr;
+    }
+
+    void* GetObject() const {
+      if (IsObject())
+        return Raw();
+      DenseLinkKey* lk = (DenseLinkKey*)Raw();
+      return lk->Raw();
+    }
+
+    // Sets pointer but preserves tagging info
+    void SetObject(void* obj) {
+      ptr_ = (void*)((uintptr_t(ptr_) & kTagMask) | (uintptr_t(obj) & ~kTagMask));
+    }
+
+    DensePtr* Next() {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+    const DensePtr* Next() const {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+   private:
+    void* ptr_ = nullptr;  //
+  };
+
+  struct DenseLinkKey : public DensePtr {
+    DensePtr next;  // could be LinkKey* or Object *.
+  };
+
+  static_assert(sizeof(DensePtr) == sizeof(uintptr_t));
+  static_assert(sizeof(DenseLinkKey) == 2 * sizeof(uintptr_t));
+
+  using LinkAllocator = std::pmr::polymorphic_allocator<DenseLinkKey>;
+  using ChainVectorIterator = std::pmr::vector<DensePtr>::iterator;
+  using ChainVectorConstIterator = std::pmr::vector<DensePtr>::const_iterator;
+
+  bool Equal(const DensePtr* dptr, const void* ptr) const {
```

Comment:
your first preference should be - to avoid indirection. Here, for example, I do not see why passing  `const DensePtr* dptr` is better than `DensePtr` by value. 
1. The latter is simpler, more readable and has fewer corner cases (you do not need to test for nullness for example). 
2. The latter is in general more efficient because compiler can use registers to pass by value. (This is less relevant here because you use an inline function , so I guess the compiler will untangle the complexity anyway, but in general, passing objects up to 16 bytes by value is most likely to be more efficient on x64 than passing them by pointer.
See https://godbolt.org/z/oeGWq4aY9  for example.